permchain/langgraph/__init__.py

```diff
@@ -0,0 +1,110 @@
+from langchain_core.runnables import Runnable, RunnableMap
+from typing import Callable, Union, Optional
+from permchain import Channel, Pregel
+
+class Actor:
+
+    def __init__(self, name: str, runnable: Runnable):
+        self.name = name
+        self.runnable = runnable
+
+
+class DecisionPoint:
+
+    def __init__(self, name: str, callable: Callable):
+        self.name = name
+        self.callable = callable
+
+class End:
+    name = "end"
+
+
+def branch(data, condition, mapping):
+    result = condition(data)
+    return Channel.write_to(mapping[result])
+
+class Graph:
+
+    def __init__(self):
+        self.nodes = {"end": End()}
+        self.connections = {}
+        self.branches = {}
+        self.entry_point: Optional[str] = None
+
+    def register(self, node: Union[Actor, DecisionPoint]):
+        if node.name in self.nodes:
+            raise ValueError(f"Actor `{node.name}` already present.")
+        self.nodes[node.name] = node
+
+    def connect(self, start: Actor, end: Union[Actor, DecisionPoint]):
+        if start.name not in self.nodes:
+            raise ValueError(f"Need to register `{start.name}` first")
+        if end.name not in self.nodes:
+            raise ValueError(f"Need to register `{end.name}` first")
+        if start.name in self.connections:
+            raise ValueError(f"Already found path for {start.name}")
+        self.connections[start.name] = end.name
+
+    def branch(self, start: DecisionPoint, end: Optional[Actor], condition: str):
```

Comment:
Would it make sense for `branch` to instead look like:

```
class Branch:
  def __init__(self, condition: str, action: Runnable):
    ...
... 

def branch(self, name: str, *branches: Branch):
  if len(branches) == 0:
    assert("ahhhhhh")
  self.branches = {}
  for branch in branches:
    self.branches[name][branch.condition] = branch
```

That way when defining your workflow you can do something like:

```
workflow.branch(
  "somebranch",
  Branch("path1", Path1Runnable),
  Branch("path2", Path2Runnable),
  Branch("stop", StopRunnable)
)
```

Reply:
If matching the proposed syntax above on `Actors`, then you'd have something more like:

```
my_decision = DecisionPoint(
  "somebranch",
   lambda x: some_fn_on_x(x),
   Branch("path1", actor1.key),
   Branch("path2", actor2.key),
   Branch("stop", actor3.key)
)
```

Then you don't even need the branch calls?